### ISSUE TITLE
5.3.0 regression - `as const` is being ignored in assignment, leading to false negatives

### ISSUE THREAD

**Issue by MichaelMitchell-at** [microsoft/TypeScript#56128]
### 🔎 Search Terms

"as const", 5.3.0, assignment

### 🕗 Version & Regression Information

- This changed between versions 5.2.2 and 5.3.0 (nightly)

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.0-dev.20231016#code/PTAEAEBcGcFoFMAeAHeBjSCBOWD2XQAVACwEtpRpjcBXAGwBNQAjeUAQ1EgE9VR4c+FunY1obAAZZ47BrgB2dbqADaAFgBMAXQmhyoebkgdo0UgHN57ZnTaRcoCQEEc7bgB55NALassAPgkAKFtjRAAuUBcsN08fP39QAF5VTS0TUDQFaEgAbiCgA

### 💻 Code

```ts
// This should be a type error because `readonly [42]` is not assignable to `Array<number>`
let x: Array<number> = [42] as const;
```

### 🙁 Actual behavior

No type error

### 🙂 Expected behavior

Type error

**Comment by Andarist** [comment]
I'm almost certain that this was changed by https://github.com/microsoft/TypeScript/pull/55229 . I think that this is the expected change in a similar vein that this is expected now: https://github.com/microsoft/TypeScript/pull/55522

**Comment by MichaelMitchell-at** [comment]
> I'm almost certain that this was changed by #55229 . I think that this is the expected change in a similar vein that this is expected now: #55522

Ah, no biggie. We were just using this behavior for a "type test", but it's easy enough to fix by assigning the argument to an intermediate variable.

**Comment by RyanCavanaugh** [comment]
@Andarist finding the PR faster than the bisect script I had going 🐇🐢!

While not identical to the issue in the linked PR, the case for this one being an error also seems a bit weak to me. You still get the desired error in the aliased case (good):
```ts
const arr = [42] as const;
let x: Array<number> = arr;
```
whereas erroring that the provably-only reference of a mutable array is an array you annotated with `as const` doesn't really seem justified. Effectively, you wrote *two* type annotations, and we picked the one that made the code work - it's not really a problem.

Aaaaaand I just saw you closed this. Cool 🥰

**Comment by Andarist** [comment]
> @Andarist finding the PR faster than the bisect script I had going 🐇🐢!

I have all Anders' PRs on my speed dial 🤣
